### PR TITLE
New article about performances, dealing with apache2 and nginx

### PR DESCRIPTION
**Pull request by djodjo3**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | New article in the performance series, dealing with apache2 vs nginx!
| Fixed ticket? | None that I'm aware of, but let me know if that's the case!

### CODE REVIEW

**Review by atomiix — COMMENTED** [review]
Be careful, I think `assets/images/2020/12/cpuusage` is a duplicate of `assets/images/2020/12/NginxVSApache2CPUUsage.png` without extension

**Comment by djodjo3** [comment]
> Be careful, I think assets/images/2020/12/cpuusage is a duplicate of assets/images/2020/12/NginxVSApache2CPUUsage.png without extension

Good catch @atomiix !
Just fixed it 👍

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+In the right corner, we have the [NGINX webserver](https://www.nginx.com/), much more recent than Apache, either under a BSD-type license (or a commercial one) and coupled with php-fpm.
+By default, and design, NGINX has an asynchronous, non-blocking event-driven architecture, meaning it does not create new threads or processes for each connection.
+
+In order to be fair, we have tested them with the exact same technical stack, the same scenario and the exact same constraints. 
+
+Here's a quick reminder of our test parameters:
+
+* We used PrestaShop version 1.7.6.3 (latest version available at the time) with the same php 7.2 configuration, in a docker environment
+* Both hosted on a not too shabby [GCP instance](https://console.cloud.google.com) with 4vCPU, 15GB RAM and SSD disks 
+* Shops are both configured with 1000 products, 10 suppliers, 2 languages, 5 categories, 1000 commands and a database around 140Mb
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
* Shops are both configured with 1000 products, 10 suppliers, 2 languages, 5 categories, 1000 orders and a database around 140Mb
```
I think you meant this?

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+![NGINXRespTime][NGINXRespTime]
+_NGINX active users are just below 100, around 94._
+{: .text-center }
+
+![ApacheRespTime][ApacheRespTime]
+_Apache active users are above 100, around 110._
+{: .text-center }
+
+Nothing suprising about the active users, given the result we already had.
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
Nothing surprising about the active users, given the result we already had. 
```

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+_NGINX active users are just below 100, around 94._
+{: .text-center }
+
+![ApacheRespTime][ApacheRespTime]
+_Apache active users are above 100, around 110._
+{: .text-center }
+
+Nothing suprising about the active users, given the result we already had. 
+
+Also, looking at the servers metrics, we see that, for both of them, we have pleny of available memory:
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
Also, looking at the servers metrics, we see that, for both of them, we have plenty of available memory:
```

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+{: .text-center }
+
+Same thing for network and disk IO, far from their limits for both instances:
+
+![DiskIO][DiskIO]
+_Nothing impressive here, especially for SSD disks - excerpt from GCP monitoring_
+{: .text-center }
+
+![NetworkTraffic][NetworkTraffic]
+_Network traffric, nothing specific, excerpt from GCP monitoring_
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
_Network traffic, nothing specific, excerpt from GCP monitoring_
```

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+![CPUUsage][CPUUsage]
+
+Some may say that looking at the CPU is not that much intesting, but others, such as us, may disagree.
+
+If you remember our previous talk about performance, you may recall that performance tuning is about finding out the current bottleneck of a system. So, if our CPU _is_ the bottleneck, then it is probably not the memory, the disk or even the network.
+
+Also meaning that the CPU can do its part properly.
+
+In our case, we can conclude there is not much constraint on the system preventing it from serving the application as expected - or at least that the constraint are similar enough.
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
In our case, we can conclude there is not much constraint on the system preventing it from serving the application as expected - or at least that the constraints are similar enough.
```

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+Also, remember that our task here is to use as much ressources as possible and CPU time means it is not waiting for IOs (disk or network), or managing RAM (neither swapping, as it should be seen on disk), and so on.
+
+
+## Finish him!
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]: Remember that Google will be crawling the HTML tags (this one will be a h2 or h3 after Markdown has been converted to HTML) and show them to everyone in Google Search Results page 😛  you might want to use a title a little more semantic...

Maybe a better title can be used and the gaming reference can be kept as a subtitle?

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+So, now that we've seen that our system is all up to work at the CPU level, let's try to understand what the Visitors result could mean to all the PrestaShops around the world.
+
+As mentioned earlier, Apache and NGINX designs are very different, Apache using threads/processes for each request and NGINX using an asynchronous system (through an event loop) to manage requests.
+
+Though NGINX's asynchronous design lets it handle enormous charges and loads that Apache can't, it has a tiny little drawback: latency.
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
Though NGINX's asynchronous design lets it handle enormous charges and loads that Apache cannot, it has a tiny little drawback: latency.
```
Suggestion to be more formal

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+This latency, at least partially, is induced by the events management - it takes time to oversee several events at the "same time", to check the different buffers statuses, and so on. Just like any other system, context switching and status management of its different parts does use resources. And more concurrent connections means more latency.
+
+Where apache reduces latency to a minimum with its synchronous design.
+
+It's also worth mentionning that the induced latency is higher for dynamic contents (such as PHP) and very low for static contents (such as JPG, PNG, JS, CSS and so on).
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
It's also worth mentioning that the induced latency is higher for dynamic contents (such as PHP) and very low for static contents (such as JPG, PNG, JS, CSS and so on).
```

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+_Garou: Mark of the Wolves belongs to SNK Corporation_
+{: .text-center }
+
+Rest assured we are aware this benchmark is not an exhaustive one and plenty of parameters could have tiped the scale one way or the other.
+
+Still, we do believe this test is relevant and could help understand which configurations are relevant to a given shop.
+
+But don't trust us blindly on this, we have provided you with [the tools](https://github.com/PrestaShop/performance-project) to perform your own benchmarks and even the capacity to enhance those tools to reach your own expectactions in different scenarios. Jut have a look and see what you can do with it.
+
+Also worth mentionning is that, using the CPU up to 98% or 99% can be fine for a benchmark but it's something you should never allow on a production server. Any additional load would slow down significantly the server and drastically increase the reponse time to a non acceptable threshold.
```
Comment (matks) [inline news/_posts/performance/2020-12-14-apache-vs-nginx.md]:
```suggestion
Also worth mentioning is that, using the CPU up to 98% or 99% can be fine for a benchmark but it's something you should never allow on a production server. Any additional load would slow down significantly the server and drastically increase the reponse time to a non acceptable threshold.
```

**Review by matks — COMMENTED** [review]
A little proof-reading and style suggestion 😉 

Also maybe we could start referencing each item of the series into one another? Pablo did it in its series about PrestaShop architecture at the end of each article https://build.prestashop.com/news/prestashop-in-2019-and-beyond-part-1-current-architecture/

This helps people, when reading the 3rd or 4th article, to go back to previous articles to refresh their memory about the topic.

**Comment by matks** [comment]
Another feedback: you and I are both French people so we know very well the western culture of gaming. But we try to make our content as accessible and as inclusive as we can. Could you add some little explanation for the choice of your images?

Suggestion:
```
This article will be illustrated by screenshots from the game (...) as this article features a fight between 2 popular webservers
```

The idea is that a Chinese reader might not understand what is going on 😄  if he never played games of this kind.

File: news/_posts/performance/2020-12-14-apache-vs-nginx.md (hunk 6 above)
Comment (djodjo3) [reply news/_posts/performance/2020-12-14-apache-vs-nginx.md]: Will work on it ;)

**Review by matks — APPROVED** [review]
Good for me 🎉

**Comment by matks** [comment]
@djodjo3 I'm OK to publish this. Have you a preferred publication date/time ?

**Comment by djodjo3** [comment]
@matks no specific schedule, we can publish whenever it suits you.
(Next article is already in the works ;) )

**Comment by matks** [comment]
Publication hour updated to 4:30pm today. When Travis validates the build, I merge.

**Comment by matks** [comment]
Thank you @djodjo3 great article !

**Comment by djodjo3** [comment]
Thank you for your patience and reviews 👍

**Pull request merged**